# Nomor 1

## Ide Dasar

1. **Import dataset** yang akan digunakan
2. **Eksplorasi dataset** untuk mengetahui bentuk, fitur, dan hal2 detail dari dataset
3. **Pre-processing** : Memilih fitur yang diperkirakan menjadi faktor penyebab heart disease
4. Membagi **data train, test,** dan **validasi**
5. **Membuat model**
6. Melakukan **hyperparameter tuning** untuk mencari parameter terbaik
7. **Membentuk model** dengan parameter yg didapat dr hyperparameter tuning
8. **Evaluasi model** dengan menge-tes nya dengan data test dan mengecek nilai accuracy, precision, recall, dan F1 score nya

## Import dataset

In [ ]:
#import package yang akan digunakan
import pandas as pd #untuk mengolah dataset
import numpy as np #untuk array dan operasi aljabar

In [ ]:
#mengimport dataset heart dlm format csv
from google.colab import files
upload = files.upload()

Saving heart.csv to heart.csv


In [ ]:
#membaca dataset heart.csv yg sudah diimport
heart_df = pd.read_csv('heart.csv')
heart_df #menampilkan data

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0     63    1   3     145   233    1        0       150     0      2.3    0   
1     37    1   2     130   250    0        1       187     0      3.5    0   
2     41    0   1     130   204    0        0       172     0      1.4    2   
3     56    1   1     120   236    0        1       178     0      0.8    2   
4     57    0   0     120   354    0        1       163     1      0.6    2   
..   ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
298   57    0   0     140   241    0        1       123     1      0.2    1   
299   45    1   3     110   264    0        1       132     0      1.2    1   
300   68    1   0     144   193    1        1       141     0      3.4    1   
301   57    1   0     130   131    0        1       115     1      1.2    1   
302   57    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output  
0      0      1       1  
1      0      2       1  
2      0      2       1  
3      0      2       1  
4      0      2       1  
..   ...    ...     ...  
298    0      3       0  
299    0      3       0  
300    2      3       0  
301    1      3       0  
302    1      2       0  

[303 rows x 14 columns]

## Eksplorasi data

In [ ]:
heart_df.info() #mengecek fitur dan jenis datanya

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trtbps    303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalachh  303 non-null    int64  
 8   exng      303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slp       303 non-null    int64  
 11  caa       303 non-null    int64  
 12  thall     303 non-null    int64  
 13  output    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [ ]:
heart_df['oldpeak'].unique() #dr df.info terlihat ada data float, sekarang lihat datanya untuk menentukan apakah perlu proses label encoding atau tidak. saya memnutuskan tidak melakukan label encoding karena menurut saya data tetap bisa diolah

array([2.3, 3.5, 1.4, 0.8, 0.6, 0.4, 1.3, 0. , 0.5, 1.6, 1.2, 0.2, 1.8,
       1. , 2.6, 1.5, 3. , 2.4, 0.1, 1.9, 4.2, 1.1, 2. , 0.7, 0.3, 0.9,
       3.6, 3.1, 3.2, 2.5, 2.2, 2.8, 3.4, 6.2, 4. , 5.6, 2.9, 2.1, 3.8,
       4.4])

In [ ]:
heart_df.columns #print semua nama fitur yang ada

Index(['age', 'sex', 'cp', 'trtbps', 'chol', 'fbs', 'restecg', 'thalachh',
       'exng', 'oldpeak', 'slp', 'caa', 'thall', 'output'],
      dtype='object')

## Pre-processing data

In [ ]:
#memilih fitur yang sekiranya berpengaruh pada kemungkinan heart disease. saya pilih 12 fitur dan 1 target

heart_new = heart_df[['age', 'cp', 'trtbps', 'chol', 'fbs', 'restecg', 'thalachh',
       'exng', 'oldpeak', 'slp', 'caa', 'thall', 'output']]
heart_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   cp        303 non-null    int64  
 2   trtbps    303 non-null    int64  
 3   chol      303 non-null    int64  
 4   fbs       303 non-null    int64  
 5   restecg   303 non-null    int64  
 6   thalachh  303 non-null    int64  
 7   exng      303 non-null    int64  
 8   oldpeak   303 non-null    float64
 9   slp       303 non-null    int64  
 10  caa       303 non-null    int64  
 11  thall     303 non-null    int64  
 12  output    303 non-null    int64  
dtypes: float64(1), int64(12)
memory usage: 30.9 KB


In [ ]:
heart_new #print dataset yg baru

age  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  caa  \
0     63   3     145   233    1        0       150     0      2.3    0    0   
1     37   2     130   250    0        1       187     0      3.5    0    0   
2     41   1     130   204    0        0       172     0      1.4    2    0   
3     56   1     120   236    0        1       178     0      0.8    2    0   
4     57   0     120   354    0        1       163     1      0.6    2    0   
..   ...  ..     ...   ...  ...      ...       ...   ...      ...  ...  ...   
298   57   0     140   241    0        1       123     1      0.2    1    0   
299   45   3     110   264    0        1       132     0      1.2    1    0   
300   68   0     144   193    1        1       141     0      3.4    1    2   
301   57   0     130   131    0        1       115     1      1.2    1    1   
302   57   1     130   236    0        0       174     0      0.0    1    1   

     thall  output  
0        1       1  
1        2       1  
2        2       1  
3        2       1  
4        2       1  
..     ...     ...  
298      3       0  
299      3       0  
300      3       0  
301      3       0  
302      2       0  

[303 rows x 13 columns]

## Train-test-validasi Data

In [ ]:
#menyimpan fitur2 berpengaruh pada array X dan target pada array Y
X = heart_new[['age', 'cp', 'trtbps', 'chol', 'fbs', 'restecg', 'thalachh',
       'exng', 'oldpeak', 'slp', 'caa', 'thall']].values
Y = heart_new['output']

 #menampilkan X dan Y
print('X :\n', X)
print('\nY :\n', Y)

X :
 [[ 63.   3. 145. ...   0.   0.   1.]
 [ 37.   2. 130. ...   0.   0.   2.]
 [ 41.   1. 130. ...   2.   0.   2.]
 ...
 [ 68.   0. 144. ...   1.   2.   3.]
 [ 57.   0. 130. ...   1.   1.   3.]
 [ 57.   1. 130. ...   1.   1.   2.]]

Y :
 0      1
1      1
2      1
3      1
4      1
      ..
298    0
299    0
300    0
301    0
302    0
Name: output, Length: 303, dtype: int64


In [ ]:
#membagi data menjadi train, test, valid data
from sklearn.model_selection import train_test_split #import package

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 3) #train : test = 8 : 2. pilih random_state = 3 agar data train dan test tidak berubah saat run kembali

#ambil data validasi sebanyak 180 < 242 (data train) dan pengambilan data 30 30
X_valid, X_trainX = X_train[:180]/30.0, X_train[180:]/30.0
y_valid, y_trainX = Y_train[:180], Y_train[180:]

#menampilkan train, test, validasi data
print('Train Set : ', X_train.shape, Y_train.shape)
print('Test Set : ', X_test.shape, Y_test.shape)
print('Validation Set : ', X_valid.shape, y_valid.shape)

Train Set :  (242, 12) (242,)
Test Set :  (61, 12) (61,)
Validation Set :  (180, 12) (180,)


## Membuat Model

In [ ]:
pip install keras #install keras

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
#import package yang akan digunakan

import keras
import tensorflow as tf

!pip install -q -U keras-tuner

import keras_tuner as kt

In [ ]:
#mendefinisikan model yang akan dibuat

def build_model(hp):
  model = keras.models.Sequential() #inisialisasi model Sequential

  hp_layers = hp.Choice('layer', [2,3]) #pemilihan layer terbaiknya 2 atau 3, karena pd soal diminta DNN jadi layer harus > 1
  hp_neurons = hp.Int('neuron', min_value = 100, max_value = 500, step = 10) #pemilihan neuron terbaik dengan range [100,500] dan stepsizenya 10
  
  for layers in range(hp_layers): #iterasi layer sesuai dengan hasil pemilihan layer terbaik
    model.add(keras.layers.Dense(hp_neurons, activation = 'tanh')) #menambahkan hidden layer sesuai jumlah layer. menggunakan fungsi aktivasi tanh karena setelah dibandingkan dgn relu, model dgn fungsi tanh lebih baik hasilnya
  model.add(keras.layers.Dense(10, activation = 'softmax')) #menampilkan outputnya

  optimizer = tf.keras.optimizers.Adam() #fungsi optimisasi menggunakan Adam karena menurut beberapa jurnal, fungsi Adam memiliki berbagai kelebihan
  model.compile(loss = 'sparse_categorical_crossentropy',
                 optimizer = optimizer,
                 metrics = ['accuracy'])
  return model #memanggil model yg sudah didefinisikan tadi

## Hyperparameter Tuning

In [ ]:
#untuk melakukan hyperparameter tuning, bentuk hyperband tuner nya dulu
tuner = kt.Hyperband(build_model,
                     objective = 'val_accuracy', #tolak ukur dari keakuratan tuningnya
                     max_epochs = 500, #maksimal epochs di setiap layer
                     directory = 'my_di', #lokasi penyimpanan file tuning
                     project_name = 'my_tuner') #lokasi penyimpanan file tuning

INFO:tensorflow:Reloading Oracle from existing project my_di/my_tuner/oracle.json
INFO:tensorflow:Reloading Tuner from my_di/my_tuner/tuner0.json


In [ ]:
#untuk memberhentikan trial jika proses trial dirasa tidak akan mendapat hasil bagus

stop_early = tf.keras.callbacks.EarlyStopping( 
    monitor = 'val_loss',
    patience = 5)

In [ ]:
#mencari hyperparameter terbaik
tuner.search(X_trainX, y_trainX, #data yg digunakan untuk melatih model dlm hyperparameter tuning
             epochs = 500, #maksimal epochs
             validation_data = [X_valid, y_valid], #data yg digunakan untuk memvalidasi data train
             callbacks = [stop_early]) #untuk memberhentikan proses trial jika dirasa trial tidak mendapat hasil signifikan

INFO:tensorflow:Oracle triggered exit


In [ ]:
best_hps = tuner.get_best_hyperparameters()[0] #menampilkan hyperparameter terbaik utnuk model ini

print("Best number of layers: ", best_hps.get("layer"))
print("Best number of neuron: ", best_hps.get("neuron"))
print("Best number of epochs: ", best_hps.get("tuner/epochs"))

Best number of layers:  3
Best number of neuron:  210
Best number of epochs:  3


In [ ]:
model = tuner.hypermodel.build(best_hps) #membentuk model dengan hyperparameter terbaik

In [ ]:
#membuat model dengan menyesuaikan dgn data train dan data validasi

model.fit(X_trainX, y_trainX,
          epochs = 3,
          validation_data = [X_valid, y_valid])

Epoch 1/3
2/2 [==============================] - 0s 78ms/step - loss: 0.6650 - accuracy: 0.5806 - val_loss: 0.6228 - val_accuracy: 0.6889
Epoch 2/3
2/2 [==============================] - 0s 48ms/step - loss: 0.6615 - accuracy: 0.6129 - val_loss: 0.6251 - val_accuracy: 0.6222
Epoch 3/3
2/2 [==============================] - 0s 48ms/step - loss: 0.6707 - accuracy: 0.5645 - val_loss: 0.6503 - val_accuracy: 0.5944


## Evaluasi Model

In [ ]:
y_proba = model.predict(X_test) #melihat probabilitas model

y_p = np.array([np.argmax(i) for i in y_proba]) #memprediksi model dengan data test
#print(y_p) untuk menampilkan hasil prediksi

[1 1 1 1 1 1 0 1 1 1 1 1 1 0 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1]


In [ ]:
#menghitung nilai akurasi, recall, presisi, dan F1 scorenya

from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score #import package yang akan digunakan

#menampilkan setiap nilainya
print('Accuracy :',(accuracy_score(Y_test, y_p)))
print('Recall:',(recall_score(Y_test, y_p)))
print('Precision:',(precision_score(Y_test, y_p)))
print('F1-Score:',(f1_score(Y_test, y_p)))

Accuracy : 0.7377049180327869
Recall: 1.0
Precision: 0.7142857142857143
F1-Score: 0.8333333333333333


**Analisis Model**

Model yang sudah dibuat adalah *Deep Neural Network* dengan 3 layer, 210 neuron, dan 3 epochs tiap layernya. Parameter tsb. dihasilkan dari proses hyperparameter tuning.


Setelah model terbentuk, dilakukan evaluasi model yaitu memprediksi Y_test dengan model yg sudah dibuat. Langkah terakhir adalah menampilkan nilai akurasi, recall, presisi, dan F1 score. Dari 4 nilai tersebut, dapat dilihat bahwa semua nilainya bernilai cukup besar, bahkan nilai recall mencapai 1. Artinya, model yg dibuat terbilang bagus dan cukup akurat.


Model ini digunakan untuk menyelesaikan masalah klasifikasi risiko penyakit jantung. Oleh karena itu, lebih baik kita menggunakan model yg memiliki nilai recall tinggi, karena nilai recall bergantung pada poin False Negative, artinya model memprediksi pasien terkena penyakit jantung padahal tidak terkena penyakit jantung. Hal ini lebih baik daripada model mendeteksi seseorang tidak terkena penyakit jantung padahal terkena penyakit jantung. 


$\therefore$ Model ini adalah model yang optimal untuk masalah klasifikasi risiko penyakit jantung, karena nilai akurasinya tinggi dan nilai recall nya sempurna, yaitu 1.0

# Nomor 2

## Ide Dasar

1. **Import dataset** yang akan digunakan
2. **Eksplorasi dataset** untuk mengetahui bentuk, fitur, dan hal2 detail dari dataset
3. **Pre-processing** untuk mengubah jenis data kategorik menjadi numerik
4. Membagi **data train, test,** dan **validasi**
5. **Membuat model**
6. Melakukan **hyperparameter tuning** untuk mencari parameter terbaik
7. **Membentuk model** dengan parameter yg didapat dr hyperparameter tuning
8. **Evaluasi model** dengan menge-tes nya dengan data test dan mengecek nilai accuracy, precision, recall, dan F1 score nya
3. **Seleksi fitur** : Memilih fitur yang diperkirakan menjadi ciri jamur beracun
9. Melakukan step 5-7 tetapi dengan dataset yg sudah melalui proses seleksi fitur
10. **Membandingkan model** yg melalui proses seleksi fitur dan tidak

## Import dataset

In [1]:
#import package yang akan digunakan
import pandas as pd #untuk mengolah dataset
import numpy as np #untuk array dan operasi aljabar

In [2]:
#mengimport dataset Jamur dlm format csv
from google.colab import files
upload = files.upload()

Saving Jamur.csv to Jamur.csv


In [3]:
#membaca dataset Jamur.csv yg sudah diimport
jamur_df = pd.read_csv("Jamur.csv")
jamur_df #menampilkan dataset yang akan digunakan

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
...    ...       ...         ...       ...     ...  ...             ...   
8119     e         k           s         n       f    n               a   
8120     e         x           s         n       f    n               a   
8121     e         f           s         n       f    n               a   
8122     p         k           y         n       f    y               f   
8123     e         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0               c         n          k  ...                        s   
1               c         b          k  ...                        s   
2               c         b          n  ...                        s   
3               c         n          n  ...                        s   
4               w         b          k  ...                        s   
...           ...       ...        ...  ...                      ...   
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                         w                      w         p          w   
1                         w                      w         p          w   
2                         w                      w         p          w   
3                         w                      w         p          w   
4                         w                      w         p          w   
...                     ...                    ...       ...        ...   
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
0              o         p                 k          s       u  
1              o         p                 n          n       g  
2              o         p                 n          n       m  
3              o         p                 k          s       u  
4              o         e                 n          a       g  
...          ...       ...               ...        ...     ...  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[8124 rows x 23 columns]

## Eksplorasi Data

In [4]:
jamur_df.info() #melihat fitur dan jenis data. terlihat semuanya bertipe object, jd harus melalui tahapan label encoding

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   class                     8124 non-null   object
 1   cap-shape                 8124 non-null   object
 2   cap-surface               8124 non-null   object
 3   cap-color                 8124 non-null   object
 4   bruises                   8124 non-null   object
 5   odor                      8124 non-null   object
 6   gill-attachment           8124 non-null   object
 7   gill-spacing              8124 non-null   object
 8   gill-size                 8124 non-null   object
 9   gill-color                8124 non-null   object
 10  stalk-shape               8124 non-null   object
 11  stalk-root                8124 non-null   object
 12  stalk-surface-above-ring  8124 non-null   object
 13  stalk-surface-below-ring  8124 non-null   object
 14  stalk-color-above-ring  

In [5]:
jamur_df.columns #melihat fitur2 nya untuk mempermudah proses label encoding

Index(['class', 'cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
       'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color',
       'stalk-shape', 'stalk-root', 'stalk-surface-above-ring',
       'stalk-surface-below-ring', 'stalk-color-above-ring',
       'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number',
       'ring-type', 'spore-print-color', 'population', 'habitat'],
      dtype='object')

## Pre-processing

In [6]:
#karena semua fiturnya masih object, jadi ubah semua fitur menjadi integer dengan Label.Encoder

from sklearn import preprocessing #import

le = preprocessing.LabelEncoder()

#label encoder semua fiturnya
jamur_df['class'] = le.fit_transform(jamur_df['class'])
jamur_df['cap-shape'] = le.fit_transform(jamur_df['cap-shape'])
jamur_df['cap-surface'] = le.fit_transform(jamur_df['cap-surface'])
jamur_df['cap-color'] = le.fit_transform(jamur_df['cap-color'])
jamur_df['bruises'] = le.fit_transform(jamur_df['bruises'])
jamur_df['odor'] = le.fit_transform(jamur_df['odor'])
jamur_df['gill-attachment'] = le.fit_transform(jamur_df['gill-attachment'])
jamur_df['gill-spacing'] = le.fit_transform(jamur_df['gill-spacing'])
jamur_df['gill-size'] = le.fit_transform(jamur_df['gill-size'])
jamur_df['gill-color'] = le.fit_transform(jamur_df['gill-color'])
jamur_df['stalk-shape'] = le.fit_transform(jamur_df['stalk-shape'])
jamur_df['stalk-root'] = le.fit_transform(jamur_df['stalk-root'])
jamur_df['stalk-surface-above-ring'] = le.fit_transform(jamur_df['stalk-surface-above-ring'])
jamur_df['stalk-surface-below-ring'] = le.fit_transform(jamur_df['stalk-surface-below-ring'])
jamur_df['stalk-color-above-ring'] = le.fit_transform(jamur_df['stalk-color-above-ring'])
jamur_df['stalk-color-below-ring'] = le.fit_transform(jamur_df['stalk-color-below-ring'])
jamur_df['veil-type'] = le.fit_transform(jamur_df['veil-type'])
jamur_df['veil-color'] = le.fit_transform(jamur_df['veil-color'])
jamur_df['ring-number'] = le.fit_transform(jamur_df['ring-number'])
jamur_df['ring-type'] = le.fit_transform(jamur_df['ring-type'])
jamur_df['spore-print-color'] = le.fit_transform(jamur_df['spore-print-color'])
jamur_df['population'] = le.fit_transform(jamur_df['population'])
jamur_df['habitat'] = le.fit_transform(jamur_df['habitat'])

In [7]:
#menampilkan dataset yang sudah melalui proses label encoding
jamur_df 

class  cap-shape  cap-surface  cap-color  bruises  odor  \
0         1          5            2          4        1     6   
1         0          5            2          9        1     0   
2         0          0            2          8        1     3   
3         1          5            3          8        1     6   
4         0          5            2          3        0     5   
...     ...        ...          ...        ...      ...   ...   
8119      0          3            2          4        0     5   
8120      0          5            2          4        0     5   
8121      0          2            2          4        0     5   
8122      1          3            3          4        0     8   
8123      0          5            2          4        0     5   

      gill-attachment  gill-spacing  gill-size  gill-color  ...  \
0                   1             0          1           4  ...   
1                   1             0          0           4  ...   
2                   1             0          0           5  ...   
3                   1             0          1           5  ...   
4                   1             1          0           4  ...   
...               ...           ...        ...         ...  ...   
8119                0             0          0          11  ...   
8120                0             0          0          11  ...   
8121                0             0          0           5  ...   
8122                1             0          1           0  ...   
8123                0             0          0          11  ...   

      stalk-surface-below-ring  stalk-color-above-ring  \
0                            2                       7   
1                            2                       7   
2                            2                       7   
3                            2                       7   
4                            2                       7   
...                        ...                     ...   
8119                         2                       5   
8120                         2                       5   
8121                         2                       5   
8122                         1                       7   
8123                         2                       5   

      stalk-color-below-ring  veil-type  veil-color  ring-number  ring-type  \
0                          7          0           2            1          4   
1                          7          0           2            1          4   
2                          7          0           2            1          4   
3                          7          0           2            1          4   
4                          7          0           2            1          0   
...                      ...        ...         ...          ...        ...   
8119                       5          0           1            1          4   
8120                       5          0           0            1          4   
8121                       5          0           1            1          4   
8122                       7          0           2            1          0   
8123                       5          0           1            1          4   

      spore-print-color  population  habitat  
0                     2           3        5  
1                     3           2        1  
2                     3           2        3  
3                     2           3        5  
4                     3           0        1  
...                 ...         ...      ...  
8119                  0           1        2  
8120                  0           4        2  
8121                  0           1        2  
8122                  7           4        2  
8123                  4           1        2  

[8124 rows x 23 columns]

## Train-test-validasi data

In [8]:
#menyimpan fitur2 pada array X dan target pada array Y
X = jamur_df[['cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
       'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color',
       'stalk-shape', 'stalk-root', 'stalk-surface-above-ring',
       'stalk-surface-below-ring', 'stalk-color-above-ring',
       'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number',
       'ring-type', 'spore-print-color', 'population', 'habitat']].values
Y = jamur_df['class']

#menampilkan X dan Y
print('X :\n', X)
print('\nY :\n', Y)

X :
 [[5 2 4 ... 2 3 5]
 [5 2 9 ... 3 2 1]
 [0 2 8 ... 3 2 3]
 ...
 [2 2 4 ... 0 1 2]
 [3 3 4 ... 7 4 2]
 [5 2 4 ... 4 1 2]]

Y :
 0       1
1       0
2       0
3       1
4       0
       ..
8119    0
8120    0
8121    0
8122    1
8123    0
Name: class, Length: 8124, dtype: int64


In [9]:
from sklearn.model_selection import train_test_split #import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 77) #membagi data menjadi data train dan test dgn perbandingan train : test = 8 : 2 dan random_state = 77 agar data train dan test tidak berubah saat run ulang

#mengambil data validasi sebanyak 4400 dari data train
X_valid, X_trainX = X_train[:4400]/220.0, X_train[4400:]/220.0
y_valid, y_trainX = Y_train[:4400], Y_train[4400:]

#menampilkan data train, test, dan validasi
print('Train Set : ', X_train.shape, Y_train.shape)
print('Test Set : ', X_test.shape, Y_test.shape)
print('Validation Set : ', X_valid.shape, y_valid.shape)

Train Set :  (6499, 22) (6499,)
Test Set :  (1625, 22) (1625,)
Validation Set :  (4400, 22) (4400,)


## Membuat Model

In [10]:
pip install keras #install keras

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [11]:
#import package yang akan digunakan untuk membuat model

import keras
import tensorflow as tf

!pip install -q -U keras-tuner

import keras_tuner as kt

     |████████████████████████████████| 133 kB 5.6 MB/s 


In [12]:
#mendefinisikan model yang akan dibuat

def build_model(hp):
  model = keras.models.Sequential() #inisialisasi model Sequential

  hp_layers = hp.Choice('layer', [3,4]) #pemilihan layer terbaiknya 3 atau 4, karena pd soal diminta DNN jadi layer harus > 1
  hp_neurons = hp.Int('neuron', min_value = 100, max_value = 500, step = 10) #pemilihan neuron terbaik dengan range [100,500] dan stepsizenya 10
  
  for layers in range(hp_layers): #iterasi layer sesuai dengan hasil pemilihan layer terbaik
    model.add(keras.layers.Dense(hp_neurons, activation = 'tanh')) #menambahkan hidden layer sesuai jumlah layer. menggunakan fungsi aktivasi tanh karena setelah dibandingkan dgn relu, model dgn fungsi tanh lebih baik hasilnya
  model.add(keras.layers.Dense(10, activation = 'softmax')) #menghasilkan output


  optimizer = tf.keras.optimizers.Adam() #fungsi optimisasi menggunakan Adam karena menurut beberapa jurnal, fungsi Adam memiliki berbagai kelebihan
  model.compile(loss = 'sparse_categorical_crossentropy',
                 optimizer = optimizer,
                 metrics = ['accuracy']) #penilaian model dengan menampilkan akurasinya
  return model #memanggil model yg sudah didefinisikan tadi

## Hyperparameter tuning

In [13]:
#bentuk hyperband sebelum melakukan hyperparameter tuning
tuner = kt.Hyperband(build_model,
                     objective = 'val_accuracy',#tolak ukur dari keakuratan tuningnya
                     max_epochs = 500, #maksimal epochs di setiap layer
                     directory = 'my_di', #lokasi penyimpanan file tuning
                     project_name = 'my_tuner') #lokasi penyimpanan file tuning

In [14]:
#untuk memberhentikan trial jika proses trial dirasa tidak akan mendapat hasil bagus

stop_early = tf.keras.callbacks.EarlyStopping(
    monitor = 'val_loss',
    patience = 5)

In [15]:
#proses pencarian hyperparameter terbaik
tuner.search(X_trainX, y_trainX, #data yg digunakan untuk melatih model dlm hyperparameter tuning
             epochs = 500, #maksimal epochs
             validation_data = [X_valid, y_valid], #data yg digunakan untuk memvalidasi data train
             callbacks = [stop_early]) #untuk memberhentikan proses trial jika dirasa trial tidak mendapat hasil signifikan

Trial 75 Complete [00h 00m 05s]
val_accuracy: 0.8670454621315002

Best val_accuracy So Far: 0.8924999833106995
Total elapsed time: 00h 05m 38s
INFO:tensorflow:Oracle triggered exit


In [16]:
#menampilkan hyperparameter terbaik untuuk model ini

best_hps = tuner.get_best_hyperparameters()[0]

print("Best number of layers: ", best_hps.get("layer"))
print("Best number of neuron: ", best_hps.get("neuron"))
print("Best number of epochs: ", best_hps.get("tuner/epochs"))

Best number of layers:  4
Best number of neuron:  460
Best number of epochs:  3


In [17]:
#membuat model dengan hasil hyperparameter tuning

model = tuner.hypermodel.build(best_hps)

In [18]:
#membuat model dengan menyesuaikan dgn data train dan data validasi

model.fit(X_trainX, y_trainX,
          epochs = 3,
          validation_data = [X_valid, y_valid])

Epoch 1/3
66/66 [==============================] - 2s 25ms/step - loss: 0.6832 - accuracy: 0.6779 - val_loss: 0.4181 - val_accuracy: 0.8409
Epoch 2/3
66/66 [==============================] - 2s 23ms/step - loss: 0.3954 - accuracy: 0.8328 - val_loss: 0.3305 - val_accuracy: 0.8730
Epoch 3/3
66/66 [==============================] - 1s 20ms/step - loss: 0.2965 - accuracy: 0.8947 - val_loss: 0.3206 - val_accuracy: 0.8914


## Evaluasi Model

In [19]:
#menghitung probabilitas model

y_proba = model.predict(X_test)

y_p = np.array([np.argmax(i) for i in y_proba]) #memprediksi model dengan data test
print(y_p) #untuk menampilkan hasil prediksi

[1 0 0 ... 1 1 0]


In [20]:
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score #import package yg digunakan untuk mencari nilai2 di bawah

#print nilai akurasi, recall, precision, F1 score
print('Accuracy :',(accuracy_score(Y_test, y_p)))
print('Recall: ',(recall_score(Y_test, y_p)))
print('Precision: ',(precision_score(Y_test, y_p)))
print('F1-Score: ',(f1_score(Y_test, y_p)))

Accuracy : 0.808
Recall: 0.6010568031704095
Precision: 0.978494623655914
F1-Score: 0.7446808510638299


## Dengan Seleksi Fitur

Untuk meningkatkan skor akurasi, akan dilakukan proses seleksi fitur

### Menyeleksi fitur

In [21]:
#definisikan jamur_fitur berisi fitur2nya dan jamur_kelas berisi kelasnya agar memudahkan proses seleksi fitur
jamur_fitur = jamur_df[['cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
       'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color',
       'stalk-shape', 'stalk-root', 'stalk-surface-above-ring',
       'stalk-surface-below-ring', 'stalk-color-above-ring',
       'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number',
       'ring-type', 'spore-print-color', 'population', 'habitat']]
jamur_kelas = jamur_df['class']

In [22]:
#menghitung korelasi (hubungan) setiap fitur2 satu sama lain

corr_matrix = jamur_fitur.corr().abs()
print(corr_matrix) #menampilkan korelasi tiap fitur

                          cap-shape  cap-surface  cap-color   bruises  \
cap-shape                  1.000000     0.050454   0.048203  0.035374   
cap-surface                0.050454     1.000000   0.019402  0.070228   
cap-color                  0.048203     0.019402   1.000000  0.000764   
bruises                    0.035374     0.070228   0.000764  1.000000   
odor                       0.021935     0.045233   0.387121  0.061825   
gill-attachment            0.078865     0.034180   0.041436  0.137359   
gill-spacing               0.013196     0.282306   0.144259  0.299473   
gill-size                  0.054050     0.208100   0.169464  0.369596   
gill-color                 0.006039     0.161017   0.084659  0.527120   
stalk-shape                0.063794     0.014123   0.456496  0.099364   
stalk-root                 0.030191     0.126245   0.321274  0.244188   
stalk-surface-above-ring   0.030417     0.089090   0.060837  0.460824   
stalk-surface-below-ring   0.032591     0.107965   

In [23]:
#karena dari hasil korelasi terlihat bahwa veil-type menghasilkan NaN, maka cek data pd fitur veil-type. ternyata fitur ini hanya berisi satu jenis data
#maka langsung hilangkan secara manual dari list fiturkarena tidak berpengaruh
jamur_df['veil-type'].unique()

array([0])

In [24]:
#jamur_fitur setelah menghapus veil-type

jamur_fitur = jamur_df[['cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
       'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color',
       'stalk-shape', 'stalk-root', 'stalk-surface-above-ring',
       'stalk-surface-below-ring', 'stalk-color-above-ring',
       'stalk-color-below-ring', 'veil-color', 'ring-number',
       'ring-type', 'spore-print-color', 'population', 'habitat']]
jamur_kelas = jamur_df['class']

In [25]:
#menghitung korelasi tiap fitur yg baru

corr_matrix = jamur_fitur.corr().abs()

In [26]:
#mengambil matriks segitiga atas

upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(np.bool))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  """Entry point for launching an IPython kernel.


In [27]:
#menghapus fitur yang kurang berkorelasi dengan fitur lain

to_drop = [column for column in upper.columns if any(upper[column] > 0.6)]
for ls in to_drop :
  jamur_fitur = jamur_fitur.drop([ls], axis = 1)

In [28]:
print(to_drop) #menampilkan fitur yang akan di hapus

['veil-color', 'ring-type', 'spore-print-color']


### Train-test-validasi data

In [29]:
from sklearn.model_selection import train_test_split #import package yg akan digunakan

X_train, X_test, Y_train, Y_test = train_test_split(jamur_fitur.values, jamur_kelas.values, test_size = 0.2, random_state = 77) #membagi train test data dari fitur yg sudah diseleksi

#mengambil 4400 data untuk dijadikan data validasi dari data train
X_valid, X_trainX = X_train[:4400]/220.0, X_train[4400:]/220.0
y_valid, y_trainX = Y_train[:4400], Y_train[4400:]

#menampilkan train-test-validasi data
print('Train Set : ', X_train.shape, Y_train.shape)
print('Test Set : ', X_test.shape, Y_test.shape)
print('Validation Set : ', X_valid.shape, y_valid.shape)

Train Set :  (6499, 18) (6499,)
Test Set :  (1625, 18) (1625,)
Validation Set :  (4400, 18) (4400,)


### Hyperparameter Tuning

In [30]:
#proses pencarian hyperparameter
tuner.search(X_trainX, y_trainX,
             epochs = 500,
             validation_data = [X_valid, y_valid],
             callbacks = [stop_early])

Trial 81 Complete [00h 00m 05s]
val_accuracy: 0.8722727298736572

Best val_accuracy So Far: 0.9061363339424133
Total elapsed time: 00h 06m 25s
INFO:tensorflow:Oracle triggered exit


In [31]:
#menampilkan hyperparameter terbaik untuuk model ini

best_hps = tuner.get_best_hyperparameters()[0]

print("Best number of layers: ", best_hps.get("layer"))
print("Best number of neuron: ", best_hps.get("neuron"))
print("Best number of epochs: ", best_hps.get("tuner/epochs"))

Best number of layers:  4
Best number of neuron:  450
Best number of epochs:  3


In [49]:
model = tuner.hypermodel.build(best_hps) #membentuk model

In [50]:
#membuat model dengan menyesuaikan dgn data train dan data validasi

model.fit(X_trainX, y_trainX,
          epochs = 3,
          validation_data = [X_valid, y_valid])

Epoch 1/3
66/66 [==============================] - 2s 22ms/step - loss: 0.7212 - accuracy: 0.6255 - val_loss: 0.4276 - val_accuracy: 0.8202
Epoch 2/3
66/66 [==============================] - 1s 21ms/step - loss: 0.3965 - accuracy: 0.8385 - val_loss: 0.3406 - val_accuracy: 0.8625
Epoch 3/3
66/66 [==============================] - 1s 19ms/step - loss: 0.3168 - accuracy: 0.8799 - val_loss: 0.3899 - val_accuracy: 0.8607


### Evaluasi Model

In [51]:
#menghitung probabilitas y

y_proba = model.predict(X_test)

y_p = np.array([np.argmax(i) for i in y_proba]) #memprediksi model dengan data test
print(y_p) #untuk menampilkan hasil prediksi

[1 0 0 ... 0 1 0]


In [52]:
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score #import sklearn.metrics untuk menghitung skor2 berikut

#menampilkan skor akurasi, recall, precision, F1 score
print('Accuracy : ',(accuracy_score(Y_test, y_p)))
print('Recall: ',(recall_score(Y_test, y_p)))
print('Precision: ',(precision_score(Y_test, y_p)))
print('F1-Score: ',(f1_score(Y_test, y_p))) 

Accuracy : 0.8246153846153846
Recall: 0.6354029062087186
Precision: 0.9816326530612245
F1-Score: 0.7714514835605453


**Analisis Model**

Setelah membentuk model didapat skor akurasi, recall, presisi, dan F1 score sebagai berikut.

***Tanpa Seleksi Fitur***

* Accuracy : 0.808
* Recall: 0.6010568031704095
* Precision: 0.978494623655914
* F1-Score: 0.7446808510638299

***Dengan Seleksi Fitur***

* Accuracy : 0.8246153846153846
* Recall: 0.6354029062087186
* Precision: 0.9816326530612245
* F1-Score: 0.7714514835605453

Terlihat, dengan melakukan proses seleksi fitur didapat nilai akurasi, recall, presisi, dan F1 scroe yang lebih tinggi.

$\therefore$ Jadi, model dengan proses seleksi fitur lebih optimal dibandingkan dengan model yang tidak melakukan seleksi fitur.